src/gmock-win32.cpp

```diff
@@ -8,73 +8,120 @@
-	bool init()
+    struct LibCore
     {
-		HMODULE module;
-		module = ::LoadLibraryA( "kernel32.dll" );
-		if ( !module ) 
-			return false;
-
-		g_pfnGetLastError = (pfnGetLastError_t)::GetProcAddress( module, "GetLastError" );
-		if ( !g_pfnGetLastError ) 
-			return false;
-		g_pfnGetCurrentProcess = (pfnGetCurrentProcess_t)::GetProcAddress( module, "GetCurrentProcess" );
-		if ( !g_pfnGetCurrentProcess ) 
-			return false;
-		g_pfnWriteProcessMemory = (pfnWriteProcessMemory_t)::GetProcAddress( module, "WriteProcessMemory" );
-		if ( !g_pfnWriteProcessMemory ) 
-			return false;
-		g_pfnVirtualProtect = (pfnVirtualProtect_t)::GetProcAddress( module, "VirtualProtect" );
-		if ( !g_pfnVirtualProtect ) 
-			return false;
-		// Doesnt matter wchar_t or char, GetModuleHandleW or GetModuleHandleA
-		g_pfnGetModuleHandleA = (pfnGetModuleHandleA_t)::GetProcAddress( module, "GetModuleHandleA" );
-		if ( !g_pfnGetModuleHandleA ) 
-			return false;
-		module = ::LoadLibraryA( "dbghelp.dll" );
-		if ( !module ) 
-			return false;
-		g_pfnImageDirectoryEntryToDataEx = (pfnImageDirectoryEntryToDataEx_t)::GetProcAddress( module, "ImageDirectoryEntryToDataEx" );
-		if ( !g_pfnImageDirectoryEntryToDataEx ) 
-			return false;
-
-		g_bInit = true;
-		return true;
-	}
-
-    LONG WINAPI InvalidReadExceptionFilter(PEXCEPTION_POINTERS /*pep*/)
+        DWORD   (WINAPI* pfn_GetLastError)(VOID){ };
+        HANDLE  (WINAPI* pfn_GetCurrentProcess)(VOID){ };
+        BOOL    (WINAPI* pfn_WriteProcessMemory)(HANDLE, LPVOID, LPCVOID, SIZE_T, SIZE_T*){ };
+        BOOL    (WINAPI* pfn_VirtualProtect)(LPVOID, SIZE_T, DWORD, PDWORD){ };
+        HMODULE (WINAPI* pfn_GetModuleHandleA)(LPCSTR){ };
+        PVOID   (WINAPI* pfn_ImageDirectoryEntryToDataEx)(PVOID, BOOLEAN, USHORT, PULONG, PIMAGE_SECTION_HEADER*){ };
+
+    public:
+        LibCore() :
+            dbghelp_ { utils::loadModule(module::dbghelp),},
+            kernel32_{ utils::loadModule(module::kernel32) }
+        {
+            setupDynLinkingFunctions();
+        }
+
+    private:
+
+#define SETUP_DYNLINK_FUNC(mod, proc) \
+    setupProcAddr(mod##_.get(), #proc, pfn_##proc); \
+    if (!pfn_##proc) \
+        THROW_ERROR_CODE(pfn_GetLastError(), "failed to setup function")
+
+#define SETUP_DYNLINK_KERNEL32_FUNC(func) SETUP_DYNLINK_FUNC(kernel32, func)
+#define SETUP_DYNLINK_DBGHELP_FUNC(func)  SETUP_DYNLINK_FUNC(dbghelp, func)
+
+        void setupDynLinkingFunctions()
+        {
+            // Kernel32
+            SETUP_DYNLINK_KERNEL32_FUNC(GetLastError);
+            SETUP_DYNLINK_KERNEL32_FUNC(GetCurrentProcess);
+            SETUP_DYNLINK_KERNEL32_FUNC(WriteProcessMemory);
+            SETUP_DYNLINK_KERNEL32_FUNC(VirtualProtect);
+            SETUP_DYNLINK_KERNEL32_FUNC(GetModuleHandleA);
+
+            // Dbghelp
+            SETUP_DYNLINK_DBGHELP_FUNC(ImageDirectoryEntryToDataEx);
+        }
+
+#undef SETUP_DYNLINK_FUNC
+#undef SETUP_DYNLINK_KERNEL32_FUNC
+#undef SETUP_DYNLINK_DBGHELP_FUNC
+
```

Comment:
Yes, many macros have been created. But this is the specificity of C++ testing, it is customary.